src/libpmemfile-core/block_array.c

```diff
@@ -0,0 +1,230 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "layout.h"
+#include "inode.h"
+#include "internal.h"
+#include "out.h"
+
+static void
+update_binfo(struct pmemfile_vinode *vinode)
+{
+	struct pmemfile_inode *inode = D_RW(vinode->inode);
+	struct block_info *binfo = &vinode->first_free_block;
+
+	if (binfo->arr != NULL) {
+		/*
+		 * If the block_info is not null, it means it was kept
+		 * up-to-date by the allocating routines.
+		 */
+		return;
+	}
+
+	/*
+	 * If binfo was not used before, it must be initialized.
+	 */
+
+	/*
+	 * Find the block_array containing the next free block metadata
+	 * slot. This is either the block_array stored right in the
+	 * inode, ...
+	 */
+	binfo->arr = &inode->file_data.blocks;
+	/*
+	 * ... or if there is more than one block_array, it is
+	 * the one linked to it with the next field.
+	 */
+	if (!TOID_IS_NULL(binfo->arr->next))
+		binfo->arr = D_RW(binfo->arr->next);
+
+	binfo->idx = 0;
+
+	/* Find the first free entry in the block array. */
+	while (binfo->idx < binfo->arr->length &&
+	    binfo->arr->blocks[binfo->idx].size != 0)
+		binfo->idx++;
+}
+
+static bool
+has_free_block_entry(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	return binfo->idx < binfo->arr->length;
+}
+
+static void
+allocate_new_block_array(struct pmemfile_vinode *vinode)
+{
+	struct pmemfile_inode *inode = D_RW(vinode->inode);
+
+	TOID(struct pmemfile_block_array) new =
+			TX_ZALLOC(struct pmemfile_block_array, FILE_PAGE_SIZE);
+	D_RW(new)->length = (uint32_t)
+			((page_rounddown(pmemobj_alloc_usable_size(new.oid)) -
+			sizeof(struct pmemfile_block_array)) /
+			sizeof(struct pmemfile_block));
+
+	D_RW(new)->next = inode->file_data.blocks.next;
+	TX_SET_DIRECT(&inode->file_data.blocks, next, new);
+	vinode->first_free_block.arr = D_RW(new);
+	vinode->first_free_block.idx = 0;
+}
+
+static struct pmemfile_block *
+acquire_new_entry(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	return binfo->arr->blocks + binfo->idx++;
+}
+
+struct pmemfile_block *
+allocate_new_block_metadata(struct pmemfile_vinode *vinode)
+{
+	update_binfo(vinode);
+
+	if (!has_free_block_entry(vinode))
+		allocate_new_block_array(vinode);
+
+	return acquire_new_entry(vinode);
+}
+
+static struct pmemfile_block *
+last_used_block(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	ASSERT(binfo->idx > 0);
+
+	return binfo->arr->blocks + (binfo->idx - 1);
+}
+
+static void
+unlink_block(struct pmemfile_block *block)
+{
+	if (!TOID_IS_NULL(block->prev)) {
+		TX_ADD_FIELD_DIRECT(D_RW(block->prev), next);
+		D_RW(block->prev)->next = block->next;
```

Comment:
Because it's useless work.

In this case I think you could replace these 2 lines with other 2 lines:
struct pmemfile_block *prev = D_RW(block->prev);
TX_SET_DIRECT(prev, next, block->next);

or even just one line:
TX_SET(block->prev, next, block->next);